### PR TITLE
feat: Addresables integtration

### PR DESCRIPTION
**Pull request by PavelKlym**
## Purpose of this PR
Very first iteration to integrate Addressables.

## Testing status
* No tests have been added.

### CODE REVIEW

File: com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs [hunk 1]
```diff
@@ -37,19 +42,29 @@ static AdditiveScenesLoader()
         /// <param name="sceneBuildIndex">Build index of he scene to be loaded.</param>
         /// <param name="loadCompleted">Load Completed callback.</param>
         /// <returns></returns>
-        public static AsyncOperation LoadAdditively(int sceneBuildIndex, Action<Scene> loadCompleted = null)
+        public static void LoadAdditively(int sceneBuildIndex, Action<Scene> loadCompleted = null)
```
Comment (stan-osipov) [inline com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs:45]: breaking chnage

Comment (stan-osipov) [reply com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs:45]: make dummy for AsyncOperation

File: com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs [hunk 2]
```diff
@@ -37,19 +42,29 @@ static AdditiveScenesLoader()
...
+            LoadAdditively(string.Empty, sceneBuildIndex, loadCompleted);
         }
 
         /// <summary>
         /// Load Scene Additively by it's name.
         /// <param name="sceneName">Name of the scene to be loaded.</param>
         /// <param name="loadCompleted">Load Completed callback.</param>
         /// </summary>
-        public static AsyncOperation LoadAdditively(string sceneName, Action<Scene> loadCompleted = null)
+        public static void LoadAdditively(string sceneName, Action<Scene> loadCompleted = null)
```
Comment (stan-osipov) [inline com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs:55]: same

File: com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs [hunk 3]
```diff
@@ -127,7 +172,17 @@ public static void Unload(Scene scene, Action unloadCompleted = null)
         /// </summary>
         public static void Unload(string sceneName, Action unloadCompleted = null)
         {
+#if UNITY_EDITOR
```
Comment (stan-osipov) [inline com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs:175]: Application.isEditor

File: com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs [hunk 4]
```diff
@@ -127,7 +172,17 @@ public static void Unload(Scene scene, Action unloadCompleted = null)
...
             Unload(sceneName, -1, unloadCompleted);
+#else
+            if (BuildConfigurationSettings.Instance.Configuration.IsSceneAddressable(sceneName))
+            {
+                UnloadAddressable(sceneName, unloadCompleted);
```
Comment (stan-osipov) [inline com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs:180]: Make sure we do not unload from cache.

Comment (stan-osipov) [reply com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs:180]: maybe we will need an additional method for that if we would want to unload addressable scene completely.

File: com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs [hunk 5]
```diff
@@ -37,19 +42,29 @@ static AdditiveScenesLoader()
...
+            LoadAdditively(sceneName, -1, loadCompleted);
+            #else
+            if (BuildConfigurationSettings.Instance.Configuration.IsSceneAddressable(sceneName))
+            {
+                LoadAddressableAdditively(sceneName, loadCompleted);
+            }
+            else {
+                LoadAdditively(sceneName, -1, loadCompleted);
+            }
+            #endif
```
Comment (stan-osipov) [inline com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs:67]: Let's separate that into models

File: com.stansassets.scene-management/package.json [hunk 6]
```diff
@@ -5,8 +5,9 @@
   "unity": "2019.3",
```
Comment (stan-osipov) [inline com.stansassets.scene-management/package.json:5]: "version": "0.0.10-preview",

File: com.stansassets.scene-management/BuildConfigurator/Editor/BuildScenesPreprocessor.cs [hunk 7]
```diff
@@ -113,5 +121,67 @@ static void RemoveScenes(ref List<string> scenes, List<SceneAsset> sceneAssets)
...
+                    continue;
+
+                var guid = AssetDatabase.AssetPathToGUID(sceneAssetPath);
+                var entry = AddressableAssetSettingsDefaultObject.Settings.CreateOrMoveEntry(guid, group, false, true);
+                entry.address = scene.name;
+            }
+        }
+    }
+
+    public static class AddressablesUtility
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Editor/BuildScenesPreprocessor.cs:169]: doc is missing.

File: com.stansassets.scene-management/BuildConfigurator/Editor/BuildScenesPreprocessor.cs [hunk 8]
```diff
@@ -113,5 +121,67 @@ static void RemoveScenes(ref List<string> scenes, List<SceneAsset> sceneAssets)
...
+    {
+        public static AddressableAssetGroup GetOrCreateGroup(string name)
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Editor/BuildScenesPreprocessor.cs:171]: doc is missing.

File: com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/AddressableSceneAssetExtension.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
+﻿using UnityEditor;
+
+namespace StansAssets.SceneManagement.Build
+{
+    public static class AddressableSceneAssetExtension
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/AddressableSceneAssetExtension.cs:5]: internal

File: com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/BuildConfigurationExtension.cs [hunk 10]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+
+namespace StansAssets.SceneManagement.Build
+{
+    public static class BuildConfigurationExtension
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/BuildConfigurationExtension.cs:7]: internal

File: com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/PlatformsConfigurationExtension.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+
+namespace StansAssets.SceneManagement.Build
+{
+    public static class PlatformsConfigurationExtension
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/PlatformsConfigurationExtension.cs:7]: internal

File: com.stansassets.scene-management/BuildConfigurator/Runtime/AddressableSceneAsset.cs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace StansAssets.SceneManagement.Build
+{
+    [Serializable]
+    public class AddressableSceneAsset
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Runtime/AddressableSceneAsset.cs:6]: doc

File: com.stansassets.scene-management/BuildConfigurator/Runtime/BuildConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using UnityEditor;
+using UnityEngine;
+
+namespace StansAssets.SceneManagement.Build
+{
+    [Serializable]
+    public class BuildConfiguration
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Runtime/BuildConfiguration.cs:10]: doc

File: com.stansassets.scene-management/BuildConfigurator/Runtime/BuildConfigurationSettings.cs [hunk 14]
```diff
@@ -6,14 +6,13 @@ namespace StansAssets.SceneManagement.Build {
     public class BuildConfigurationSettings :  PackageScriptableSettingsSingleton<BuildConfigurationSettings>
     {
```
Comment (stan-osipov) [inline com.stansassets.scene-management/BuildConfigurator/Runtime/BuildConfigurationSettings.cs:7]: doc

File: com.stansassets.scene-management/package.json (hunk 6 above)
Comment (stan-osipov) [reply com.stansassets.scene-management/package.json:5]: "unity": "2018.4",

**Review by stan-osipov — CHANGES_REQUESTED** [review]
Awesome! Some changes requested.

File: com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/PlatformsConfigurationExtension.cs (hunk 11 above)
Comment (PavelKlym) [reply com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/PlatformsConfigurationExtension.cs]: fixed

File: com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/BuildConfigurationExtension.cs (hunk 10 above)
Comment (PavelKlym) [reply com.stansassets.scene-management/BuildConfigurator/Editor/Extensions/BuildConfigurationExtension.cs]: fixed

File: com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs (hunk 3 above)
Comment (PavelKlym) [reply com.stansassets.scene-management/Runtime/Statis/AdditiveScenesLoader.cs]: fixed

File: com.stansassets.scene-management/BuildConfigurator/Runtime/BuildConfiguration.cs [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+                return asset.Addressable;
+            }
+            return false;
+        }
+
+        Dictionary<string, AddressableSceneAsset> m_AddressableSceneNamesToSceneAssets;
+        [SerializeField] List<string> m_SceneNames = new List<string>();
+        [SerializeField] List<AddressableSceneAsset> m_SceneAssets = new List<AddressableSceneAsset>();
+        #if UNITY_EDITOR
+        public void InitializeBuildData(BuildTargetRuntime builtTarget)
```
[the remark this answers is not in the document]
Comment (PavelKlym) [reply com.stansassets.scene-management/BuildConfigurator/Runtime/BuildConfiguration.cs]: fixed

**Pull request merged**